Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Cancel.hs

```diff
@@ -157,26 +163,69 @@ cancel req merchant booking mbActiveSearchTry = do
+
+customerCancellationChargesCalculation :: SRB.Booking -> Maybe SRide.Ride -> Flow (Maybe PriceAPIEntity)
+customerCancellationChargesCalculation booking mbRide = do
+  transporterConfig <- CCT.findByMerchantOpCityId booking.merchantOperatingCityId (Just (TransactionId (Id booking.transactionId))) >>= fromMaybeM (TransporterConfigNotFound booking.merchantOperatingCityId.getId)
+  if not transporterConfig.canAddCancellationFee
+    then return Nothing
+    else do
+      mbCancellationAndNoShowConfigs <- CQCancellationAndNoShowChargeConfigs.findByMerchantOpCityIdAndServiceTierType booking.merchantOperatingCityId booking.vehicleServiceTier
+      case (mbRide, mbCancellationAndNoShowConfigs) of
+        (Just ride, Just cancellationAndNoShowConfigs) -> do
+          now <- getCurrentTime
+          currDistanceToPickup <- getDistanceToPickup ride
+          let initialDistanceToPickup = booking.distanceToPickup
+              timeSpentByDriver = roundToIntegral $ diffUTCTime now ride.createdAt
+              isTimeSpentByDriverLessThanThreshold = timeSpentByDriver <= cancellationAndNoShowConfigs.freeCancellationTimeSeconds
+          driverDetails <- QDI.findById ride.driverId >>= fromMaybeM (PersonNotFound ride.driverId.getId)
+          if isTimeSpentByDriverLessThanThreshold || driverDetails.hasAdvanceBooking
+            then return Nothing
+            else do
+              cancellationCharges <- case (currDistanceToPickup, initialDistanceToPickup) of
+                (Just currDist, Just initDist) -> do
+                  let distanceTravelledByDriver = initDist - currDist
+                      distanceCharges = cancellationAndNoShowConfigs.perMetreCancellationCharge * (toHighPrecMoney distanceTravelledByDriver.getMeters)
+                      timeCharges = toHighPrecMoney (timeSpentByDriver.getSeconds `div` 60) * cancellationAndNoShowConfigs.perMinuteCancellationCharge
+                      percentageOfRideFare = (cancellationAndNoShowConfigs.percentageOfRideFareToBeCharged * booking.estimatedFare) / 100
+                      timeAndDistanceCharges = distanceCharges + timeCharges
+                      minCharge = cancellationAndNoShowConfigs.minCancellationCharge
+                      maxCharge = cancellationAndNoShowConfigs.maxCancellationCharge
+                      cancellationFee = max minCharge (min timeAndDistanceCharges (min percentageOfRideFare maxCharge))
+                  return cancellationFee
+                _ -> return cancellationAndNoShowConfigs.minCancellationCharge
+              QRide.updateCancellationFeeIfCancelledField (Just cancellationCharges) ride.id
+              return $ Just $ PriceAPIEntity {amount = cancellationCharges, currency = cancellationAndNoShowConfigs.currency}
+        _ -> do
+          logError "customerCancellationChargesCalculation: Ride or cancellation and now show configs not found"
+          return Nothing
+  where
+    getDistanceToPickup ride = do
```

Comment:
We are doing the same thing in `cancel` function too? Do we need to repeat it? Can't it be reused?

Reply:
Yes, made change